adapters/qt/qt.go

```diff
@@ -0,0 +1,158 @@
+package qt
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+
+	"github.com/prebid/openrtb/v20/openrtb2"
+	"github.com/prebid/prebid-server/v2/adapters"
+	"github.com/prebid/prebid-server/v2/config"
+	"github.com/prebid/prebid-server/v2/openrtb_ext"
+)
+
+type adapter struct {
+	endpoint string
+}
+
+type reqBodyExt struct {
+	QTBidderExt reqBodyExtBidder `json:"bidder"`
+}
+
+type reqBodyExtBidder struct {
+	Type        string `json:"type"`
+	PlacementID string `json:"placementId,omitempty"`
+	EndpointID  string `json:"endpointId,omitempty"`
+}
+
+func Builder(bidderName openrtb_ext.BidderName, config config.Adapter, server config.Server) (adapters.Bidder, error) {
+	bidder := &adapter{
+		endpoint: config.Endpoint,
+	}
+	return bidder, nil
+}
+
+func (a *adapter) MakeRequests(request *openrtb2.BidRequest, reqInfo *adapters.ExtraRequestInfo) ([]*adapters.RequestData, []error) {
+	var errs []error
+	var adapterRequests []*adapters.RequestData
+
+	reqCopy := *request
+	for _, imp := range request.Imp {
+		reqCopy.Imp = []openrtb2.Imp{imp}
+
+		var bidderExt adapters.ExtImpBidder
+		var qtExt openrtb_ext.ImpExtQT
+
+		if err := json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+		if err := json.Unmarshal(bidderExt.Bidder, &qtExt); err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		impExt := reqBodyExt{QTBidderExt: reqBodyExtBidder{}}
+
+		if qtExt.PlacementID != "" {
+			impExt.QTBidderExt.PlacementID = qtExt.PlacementID
+			impExt.QTBidderExt.Type = "publisher"
+		} else if qtExt.EndpointID != "" {
+			impExt.QTBidderExt.EndpointID = qtExt.EndpointID
+			impExt.QTBidderExt.Type = "network"
+		}
+
+		finalyImpExt, err := json.Marshal(impExt)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		reqCopy.Imp[0].Ext = finalyImpExt
+
+		adapterReq, err := a.makeRequest(&reqCopy)
+		if err != nil {
+			errs = append(errs, err)
+			continue
+		}
+
+		adapterRequests = append(adapterRequests, adapterReq)
+	}
+
+	if len(adapterRequests) == 0 {
+		errs = append(errs, errors.New("found no valid impressions"))
+		return nil, errs
+	}
+
+	return adapterRequests, nil
```

Comment:
These could be combined as - `return adapterRequests, append(errs, errors.New("found no valid impressions"))`

Reply:
@qt-io please update on the comment above.

Reply:
@qt-io Gentle reminder.

Reply:
@gargcreation1992 but this code will append error "found no valid impressions" even if len(adapterRequests) > 0
So, what's the purpose of that? Perhaps I'm missing something.

Reply:
^ Looks good to me. These are equivalent since adapterRequests will be nil if never appended.

Not sure what value a "found no valid impressions" error provides since there must be at least one error for this condition to be hit.